README.md
# gratipay-badge

GitHub styled badge for [Gratipay][]

PNG: ![Gratipay badge][] SVG: ![Gratipay badge][svg-badge]

[Gratipay]: https://gratipay.com/
[Gratipay badge]: dist/gratipay.png
[svg-badge]: http://rawgit.com/gratipay/gratipay-badge/master/dist/gratipay.svg

## Usage
To ensure you are using a stable badge, it is suggested you use a semver'd badge.

### Markdown
**PNG:**

```md
[![Support via Gratipay](https://cdn.rawgit.com/gratipay/gratipay-badge/2.3.0/dist/gratipay.png)](https://gratipay.com/USERNAME/)
```

**SVG:**

```md
[![Support via Gratipay](https://cdn.rawgit.com/gratipay/gratipay-badge/2.3.0/dist/gratipay.svg)](https://gratipay.com/USERNAME/)
```

### HTML
**PNG:**

```html
<a href="https://gratipay.com/USERNAME/">
  <img alt="Support via Gratipay" src="https://cdn.rawgit.com/gratipay/gratipay-badge/2.3.0/dist/gratipay.png"/>
</a>
```

**SVG:**

```html
<a href="https://gratipay.com/USERNAME/">
  <img alt="Support via Gratipay" src="https://cdn.rawgit.com/gratipay/gratipay-badge/2.3.0/dist/gratipay.svg"/>
</a>
```

### Raw URL
**PNG:**

```
https://cdn.rawgit.com/gratipay/gratipay-badge/2.3.0/dist/gratipay.png
```

**SVG:**

```
https://cdn.rawgit.com/gratipay/gratipay-badge/2.3.0/dist/gratipay.svg
```

### Minor and patch releases
As of `2.0.0`, we offer variable minor and patch branches. For example, if you would like `gratipay-badge@2` but would like backwards compatible changes in the future, then you can use `2.x.x` over `2.3.0`.

**PNG:**

```
// Allow minor updates
https://cdn.rawgit.com/gratipay/gratipay-badge/2.x.x/dist/gratipay.png

// Allow patch updates
https://cdn.rawgit.com/gratipay/gratipay-badge/2.3.x/dist/gratipay.png
```

**SVG:**

```
https://cdn.rawgit.com/gratipay/gratipay-badge/2.x.x/dist/gratipay.svg
https://cdn.rawgit.com/gratipay/gratipay-badge/2.3.x/dist/gratipay.svg
```

## Documentation
The latest images are located under `dist/gratipay.png` and `dist/gratipay.svg`. If you would like to build your own, run `./build.sh`.

You must have [PhantomJS][] and [pngcrush][] installed for the build script to work. We require [pngcrush][] since the output from [PhantomJS][] is approximately 12x larger.

If you would like to adjust the image, you can find the HTML reference page in `lib/index.html`. It is suggested that you use [serve][] and [livereload][] to host and develop against the image.

[PhantomJS]: http://phantomjs.org/
[pngcrush]: http://pmt.sourceforge.net/pngcrush/
[serve]: https://npmjs.org/package/serve
[livereload]: https://github.com/lepture/python-livereload

## Donating
Support this project and [others by twolfson][gratipay-twolfson] via [gratipay][gratipay-twolfson].

[![Support via Gratipay][gratipay]][gratipay-twolfson]

[gratipay]: https://cdn.rawgit.com/gratipay/gratipay-badge/2.3.0/dist/gratipay.png
[gratipay-twolfson]: https://gratipay.com/twolfson/

## License
As of Feb 19 2014, Todd Wolfson has released this repository and its contents to the public domain.

It has been released under the [UNLICENSE][].

[UNLICENSE]: UNLICENSE

Previous to this, it was licensed under the [MIT license][].

[MIT license]: https://github.com/gratipay/gratipay-badge/blob/ee5cc0ad6573ef6e80048c7229bc1b7c01942b4c/README.md#license


CHANGELOG.md
# gratipay-badge changelog
2.3.0 - Automated SVG properties to be built from HTML

2.2.0 - Added SVG image via @lvillani in #9

2.1.5 - Upgraded `semver-branches` to fix keyword release issue

2.1.4 - Added `foundry-release-git-semver-branches` to manage semver variable branches (e.g. `2.x.x` `2.3.x`)

2.1.3 - Added `foundry` and plugins to make releasing easier

2.1.2 - Added `CONTRIBUTING.md`

2.1.1 - Created `update-semver.sh` to make semver updates easier

2.1.0 - Updated `gratipay` URL to be `www.` free. Via @whit537 in #8

2.0.1 - Added documentation for minor branches, patch branches, and releasing

2.0.0 - Transferred to `gratipay` organization and renamed to `gratipay-badge`

1.1.2 - Re-merged missing changes from 1.1.1

1.1.1 - Robustified build script with echoes, early exits, and typos for filenames

1.1.0 - Updated README to point to `cdn.rawgit.com` over `rawgithub.com`

1.0.0 - Moved button content from "Gittip" to "Gratipay"

0.2.1 - Moved from MIT license to UNLICENSE

0.2.0 - Moved to automated build via `html2canvas`

0.1.2 - Correcting spelling error

0.1.1 - Updated donation section

0.1.0 - Initial release


update-semver.sh
#!/bin/bash
# Helper script for update semver in README and other docs
# Echo command and exit on first error
set -e
set -x

# Alias `sort` for consistency across platforms
if which gsort > /dev/null; then
  alias sort="gsort"
fi

# If there is no semver provided, complain and leave
semver="$1"
if test "$semver" = ""; then
  echo "No \`$semver\` was provided to \`update-semver.sh\`. Please provide one (e.g. \`./update-semver.sh 1.2.3\`" 1>&2
  exit 1
fi

# Grab the last semver and its branches (1.2.3 -> 1.x.x, 1.2.x)
last_semver="$(git tag | sort -V | tail -n 1)"
last_minor_branch="$(echo $last_semver | sed -r "s/([0-9]+)\.([0-9]+)\.[0-9]+/\1.x.x/")"
last_patch_branch="$(echo $last_semver  | sed -r "s/([0-9]+)\.([0-9]+)\.[0-9]+/\1.\2.x/")"

# Calculate the new branches
minor_branch="$(echo $semver | sed -r "s/([0-9]+)\.([0-9]+)\.[0-9]+/\1.x.x/")"
patch_branch="$(echo $semver  | sed -r "s/([0-9]+)\.([0-9]+)\.[0-9]+/\1.\2.x/")"

# Replace all instances in the README
sed --in-place "s/$last_semver/$semver/g" README.md
sed --in-place "s/$last_minor_branch/$minor_branch/g" README.md
sed --in-place "s/$last_patch_branch/$patch_branch/g" README.md


lib/gittip-badge.js
var fs = require('fs');
var page = require('webpage').create();

// Grabbed from https://github.com/twolfson/icomoon-phantomjs/blob/0.2.7/lib/icomoon-phantomjs.js#L5-L14
// Run a check function every 100 ms
function waitFor(checkFn, cb) {
  if (checkFn()) {
    cb();
  } else {
    setTimeout(function waitForFn () {
      waitFor(checkFn, cb);
    }, 100);
  }
}

// Taken from https://gist.github.com/atk/1020396
var decodeBase64 = function(d,b,c,u,r,q,x){for(r=q=x='';c=d[x++];~c&&(u=q%4?u*64+c:c,q++%4)?r+=String.fromCharCode(255&u>>(-2*q&6)):0)c=b.indexOf(c);return r};
var base64Map = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';

// Open our local file
var __dirname = phantom.libraryPath;
page.open(__dirname + '/index.html', function (status) {
  // If it failed to open, throw a tantrum
  if (status === 'fail') {
    throw new Error('Could not open "index.html"');
  }

  // Get the rendered image
  function getImageStr() {
    return page.evaluate(function () {
      return window.image;
    });
  }
  waitFor(function waitForImageToLoad () {
    return getImageStr();
  }, function handleImage () {
    // Convert 'data:image/png;base64,iVBORw0KGgoAAAANSU' into an image on disk
    var imageUri = getImageStr();
    var imageStr = imageUri.replace('data:image/png;base64,', '');
    var binaryImage = decodeBase64(imageStr, base64Map);
    var destPath = __dirname + '/../dist/gittip.png';
    fs.write(destPath, binaryImage, 'wb');
    console.log('Wrote gittip badge to "' + destPath + '"');
    phantom.exit(0);
  });
});

lib/gratipay-badge.js
var fs = require('fs');
var page = require('webpage').create();

// Grabbed from https://github.com/twolfson/icomoon-phantomjs/blob/0.2.7/lib/icomoon-phantomjs.js#L5-L14
// Run a check function every 100 ms
function waitFor(checkFn, cb) {
  if (checkFn()) {
    cb();
  } else {
    setTimeout(function waitForFn () {
      waitFor(checkFn, cb);
    }, 100);
  }
}

// Taken from https://gist.github.com/atk/1020396
var decodeBase64 = function(d,b,c,u,r,q,x){for(r=q=x='';c=d[x++];~c&&(u=q%4?u*64+c:c,q++%4)?r+=String.fromCharCode(255&u>>(-2*q&6)):0)c=b.indexOf(c);return r};
var base64Map = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';

// Open our local file
var __dirname = phantom.libraryPath;
page.open(__dirname + '/index.html', function (status) {
  // If it failed to open, throw a tantrum
  if (status === 'fail') {
    throw new Error('Could not open "index.html"');
  }

  // Get the rendered image
  function getImageStrs() {
    return page.evaluate(function () {
      if (window.pngImage && window.svgImage) {
        return {
          png: window.pngImage,
          svg: window.svgImage
        };
      }
    });
  }
  waitFor(function waitForImageToLoad () {
    return getImageStrs();
  }, function handleImage () {
    // Load PNG and SVG data
    var imageUris = getImageStrs();

    // Convert 'data:image/png;base64,iVBORw0KGgoAAAANSU' into an image on disk
    var pngUri = imageUris.png;
    var pngStr = pngUri.replace('data:image/png;base64,', '');
    var binaryPng = decodeBase64(pngStr, base64Map);
    var destPngPath = __dirname + '/../dist/gratipay.png';
    fs.write(destPngPath, binaryPng, 'wb');
    console.log('Wrote PNG gratipay badge to "' + destPngPath + '"');

    // Output SVG badge
    var svgStr = imageUris.svg;
    var destSvgPath = __dirname + '/../dist/gratipay.svg';
    fs.write(destSvgPath, svgStr, 'w');
    console.log('Wrote SVG gratipay badge to "' + destSvgPath + '"');

    // Leave successfully
    phantom.exit(0);
  });
});
